### PR TITLE
rubocop: Only enable `Style/Documentation` for `@api public` code

### PR DESCRIPTION
**Pull request by issyl0**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

- Suggested in https://github.com/Homebrew/brew/pull/14709#issuecomment-1437461642.
- Found the public API paths with `git grep -l "@api public"`.

### CODE REVIEW

File: Library/Homebrew/utils.rb [hunk 1]
```diff
@@ -84,7 +84,7 @@ def inject_dump_stats!(the_module, pattern)
   # rubocop:enable Style/GlobalVars
 end
 
-module Kernel
+module Kernel # rubocop:disable Style/Documentation
```
Comment (issyl0) [inline Library/Homebrew/utils.rb:87]:
```
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/utils.rb:87:15: W: [Correctable] Lint/RedundantCopDisableDirective: Unnecessary disabling of Style/Documentation.
[11](https://github.com/Homebrew/brew/actions/runs/4275555139/jobs/7442995057#step:6:12)
module Kernel # rubocop:disable Style/Documentation
[12](https://github.com/Homebrew/brew/actions/runs/4275555139/jobs/7442995057#step:6:13)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

❓

Comment (issyl0) [reply Library/Homebrew/utils.rb:87]: I think this is the "invoking `brew style` from `/opt/homebrew` or `/opt/homebrew/Library/Homebrew` issue again.

Comment (issyl0) [reply Library/Homebrew/utils.rb:87]: The easiest way to fix this is to add a comment to `module Kernel` in `Homebrew/utils.rb` and then the test passes and the linter is happy. No more in-line excludes.

File: Library/.rubocop.yml [hunk 2]
```diff
@@ -339,13 +339,47 @@ Style/DisableCopsWithinSourceCodeDirective:
...
 Style/Documentation:
-  Exclude:
-    - "Taps/**/*"
-    - "/**/{Formula,Casks}/*.rb"
-    - "**/{Formula,Casks}/*.rb"
-    - "**/*.rbi"
+  AllowedConstants:
+    - Homebrew
+  Include:
+    - Homebrew/cask/dsl.rb
```
Comment (MikeMcQuaid) [inline Library/.rubocop.yml:347]: Feels like this should probably be in the `Homebrew/.rubocop.yml` considering it's scoped entirely to there? Thoughts?

**Review by MikeMcQuaid — COMMENTED** [review]
This is nice! Would be also cool if we could somehow detect if private APIs are used in `brew style` one day (not in this PR!) so we can warn about them (CC @dduugg for @Bo98 if you have any ideas).

> Found the public API paths with git grep -l "@api public".

The only thing I wonder here is if it'd be nice to turn the above, in this PR, into a CI check so stop this list getting out of sync?

File: Library/.rubocop.yml [hunk 3]
```diff
@@ -339,14 +339,6 @@ Style/DisableCopsWithinSourceCodeDirective:
     - "/**/{Formula,Casks}/*.rb"
     - "**/{Formula,Casks}/*.rb"
 
-# Don't enforce documentation in casks or formulae.
```
Comment (MikeMcQuaid) [inline Library/.rubocop.yml:342]: Still need this here, I think, or to disable it or it'll trigger for taps?

File: Library/Homebrew/utils.rb [hunk 4]
```diff
@@ -84,6 +84,7 @@ def inject_dump_stats!(the_module, pattern)
   # rubocop:enable Style/GlobalVars
 end
 
+# Contains shorthand Homebrew utility methods like `ohai`, `opoo`, `odisabled`.
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils.rb:87]: 🎉

File: Library/.rubocop.yml (hunk 3 above)
Comment (issyl0) [reply Library/.rubocop.yml:342]: Oh, yeah. The separate files thing is confusing.

**Comment by issyl0** [comment]
> The only thing I wonder here is if it'd be nice to turn the above, in this PR, into a CI check so stop this list getting out of sync?

Done!

File: .github/workflows/rubocop-filepaths.yml [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+jobs:
+  check:
+    runs-on: ubuntu-22.04
+    if: github.repository == 'Homebrew/brew'
+    steps:
+      - name: Checkout repo
+        uses: actions/checkout@v3
+
+      - name: Check public API docs
```
Comment (MikeMcQuaid) [inline .github/workflows/rubocop-filepaths.yml:17]: Great work! Given this is so quick after checkout (the slowest part): please add to the existing `syntax` job instead, thanks!

File: Library/Homebrew/.rubocop.yml [hunk 6]
```diff
@@ -17,27 +17,47 @@ Naming/PredicateName:
     - is_32_bit?
     - is_64_bit?
 
+# Only enforce documentation for public APIs.
```
Comment (MikeMcQuaid) [inline Library/Homebrew/.rubocop.yml:20]:
```suggestion
# Only enforce documentation for public APIs.
# Checked by the tests.yml syntax job
```

**Review by MikeMcQuaid — APPROVED** [review]

File: .github/workflows/tests.yml [hunk 7]
```diff
@@ -46,6 +46,19 @@ jobs:
 
       - run: brew typecheck
 
+      - name: Check RuboCop filepaths
+        working-directory: Library/Homebrew
```
Comment (MikeMcQuaid) [inline .github/workflows/tests.yml:50]: This doesn't exist? https://github.com/Homebrew/brew/actions/runs/4293152654/jobs/7480482506#step:8:19

Comment (issyl0) [reply .github/workflows/tests.yml:50]: 🤔 It existed with the previous `actions/checkout` step, I wonder what `setup-homebrew` is doing?

Comment (MikeMcQuaid) [reply .github/workflows/tests.yml:50]: @issyl0 Maybe just running in another directory?

Comment (issyl0) [reply .github/workflows/tests.yml]: Fixed!

**Pull request merged**